### ISSUE TITLE
Use slf4j for logging

### ISSUE THREAD

**Issue by kilink** [facebook/facebook-java-business-sdk#29]
The package currently uses a PrintStream instance for logging. IMO it would be better to use slf4j, which gives consumers control over which logging framework is used, how logging is done, finer grained control over which kinds of messages are logged, etc.

**Comment by JiamingFB** [comment]
Currently the logging is only for debugging purpose.
You can actually simply override print() in PrintStream, and redirect the output to slf4j, then you can use whichever framework you want.

**Comment by kilink** [comment]
> Currently the logging is only for debugging purpose.

Isn't that what most logging is for?

> You can actually simply override print() in PrintStream, and redirect the output to slf4j, then you can use whichever framework you want.

Okay, but that's the whole point of slf4j; it's simply a facade which lets the consumer supply the logging framework of her choice.

Using slf4j has other advantages:

- You don't need a separate debug flag.
- The consumer can have finer grained control over what to log and when. For example, while debugging an issue, one might log everything, but in steady state just log errors to avoid polluting the logs.
- String formatting support in slf4j and similar logging frameworks avoid allocating unnecessary objects when the message is simply discarded. Here is an [example](https://github.com/facebook/facebook-java-ads-sdk/blob/69d576ad144e2e320907160c7088e6bbfcaf4a29/src/main/java/com/facebook/ads/sdk/AdAccountUser.java#L81-L82). Granted, it may not be significant in most cases, but some people like to avoid unnecessary garbage.
- Customization of the message format. The consumer can choose to log timestamps, and other auxiliary information.
- No need for special configuration (like your suggestion to create a PrintStream class). A consumer's other dependencies will typically have their logging configured in one place.

Maybe you think it's overkill, or don't want to add the dependency? I can already see a need for separate log levels, since as is there is only a single flag, so you either get a huge amount of information, or none at all. Additionally, as the library grows, so will the need to log different kinds of information (exceptions, warnings, etc). And regarding adding a dependency, I am going to guess that the vast majority of consumers already depend on slf4j, either directly or transitively. If this is a serious impediment, then I think even using java.util.logging would be a better option.

To be clear, I'm not asking anyone to do this work. I'll gladly contribute the PR for this (and any other issues I've opened). First I just want to figure out whether it will be accepted or not :-)

**Comment by JiamingFB** [comment]
Hi kilink,

I actually agree that the logging needs improvement. Current logging is only for SDK debugging to check if the HTTP request is assembled and sent correctly, so it's not well structured. I think many people do want to log and archive all the requests they send in a better structure (maybe a JSON string like {"request": {...}, "response": {...}} ?), which is not supported right now but would be great to have, and in this case slf4j will be a good choice. I think slf4j is lightweight so I'm not worried about dependency.

This project is open-source and we're happy to see people contribute to it. You're more than welcome to make contributions :)

Thanks!

**Comment by nemccarthy** [comment]
+1

**Comment by codytwinton** [comment]
It seems like this issue has been open for more than a few months now. We have made many changes in the SDK since then, including releasing a new version. If you can repro your specific issue with the latest version of the SDK, please create a new issue or comment here with further details.
